### PR TITLE
Fix issue 20421 - UWP Xbox Python add-ons permission error

### LINKED ISSUE #20421: Kodi 19.3 Matrix Xbox - Python add-ons error's
Kodi 19.3 Matrix
Xbox Series S

With one the latest update Kodi 19 matrix on the Xbox consoles I am unable to load any add-ons program them or anything when trying to program one it just says the wizard build name error check the log for more information there's nothing in the log redownloading app didn't help

### PR DESCRIPTION
**Pull request by Chicopower**
## Description
With Kodi Matrix 19.3 on Xbox in retail mode, the external add-ons may get "Permission denied" error related to __init__.py files.
The problem appear when Kodi Python add-on script present in User folder (updated add-on OR a not built-in add-on) depend on a Kodi Python library present in Application Folder (built-in Python library packaged with Kodi release).
In Kodi UWP compilation process, this fix will add a hash comment character in empty __init__.py Python Lib files.
This fix will only run for the Xbox platform compilation.

## Motivation and context
In love with Kodi and free Open Source projects.
Here, i want to fix the current most critical Kodi Xbox issue safely.
See more info here: https://forum.kodi.tv/showthread.php?tid=365157

## How has this been tested?
Using my Xbox Series X with Kodi 19.3 matrix branch code in Retail mode (cannot be reproduced in dev mode).
Was able to reproduce the issue with my "private Kodi 19.3 release" in Retail mode.
After that, i was able to introduce and test this fix on my "private Kodi 19.3 release" in Retail mode.
I tested with Youtube add-on which require "urllib" built-in Kodi Python library and this fix work fine (fix the issue).
See open source code with the included fix here (this is my "private Kodi 19.3 release" source): https://github.com/Chicopower/xbmc/tree/XBMCBetaTesting-Matrix

## What is the effect on users?
With this fix, Kodi Xbox users will be able to use their external Python add-ons without "Permission denied" add-on crash related to __init__.py files.

## Types of change

- [ x] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
- [ x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x ] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

**Comment by fuzzard** [comment]
I would look to move this into an install.cmake file located in https://github.com/xbmc/xbmc/tree/master/cmake/scripts/windowsstore

You can check osx as an example, but essentially its used for packaging steps that are platform specific.

it gets called at https://github.com/xbmc/xbmc/blob/61becb5c50be3fd79bfd9a67da14d8f60fb5deba/CMakeLists.txt#L449

**Comment by matthuisman** [comment]
@Chicopower
Any research done into what is actually causing the issue? We know the workaround, but what's the cause?

**Review by wsnipex — CHANGES_REQUESTED** [review]
please move to an uwp specific file as suggested by fuzzard

**Comment by DaveTBlake** [comment]
@Chicopower thank you for your interest and work on this issue. 

>I assumed a team member may have contact with a Microsoft human and be able to find out what's going on.

@matthuisman if only it was that easy. Yes it would be nice to know why the published app has this problem while it doesn't when side loaded. But we may well go ahead with a workaround if there is no one to take such investigation futher.

**Comment by afedchin** [comment]
@Chicopower It looks like you are trying to cure symptoms instead of reasons. 
I'm almost sure that the core of the issue in python library's patch for uwp. perhaps there is an issue with processing zero-length files in the patch. 
There wasn't the issue with py2 and Kodi18.x

**Comment by thexai** [comment]
Are there files of zero bytes in 18.x / PY2?

**Comment by Chicopower** [comment]
Hi,thx @fuzzard for the hint. I applied your recommandation. The uwp package seem to be also fixed with the code moved there. I now clearly see my cmake log too, nice! Thx! 👍 

@afedchin 
As i said in the description, empty init.py files is not always a problem for UWP Kodi Python 3 scripts.
It's a UWP problem if a script present in User Folder (Q:\Users\...\AppData\) need to use a dependency present in Application Folder (S:\Program Files\WindowsApps\...) in Retail mode. The Python dependency mechanism will read the Python Lib __init__.py files to determine the available dependencies. If this file is empty, we have the permission denied error when executed in Retail mode.

The other way we can fix this issue is at Application startup. We can add a step in add-on migration (executed after "database migration") to copy the new Kodi Python built-in Libraries in a "Kodi version specific Python Libraries folder" under User folder. Then, for Kodi Xbox users, we will define PYTHONPATH environnement to use this "Kodi version specific Python Libraries folder" present in User folder instead of the one present in Application Folder.
I prefer the current way to fix it.
It appear to be more simple.
If this pull request is refused, i can work on this other possible solution.


@matthuisman Thx for your contribution.
What was required for this solution is :
- Investigate the issue (see description)
- Setup and publish a Beta version in Microsoft Store with Microsoft approvals
- Add debug infos for issue investigation
- Test without and with some changes using this Kodi Xbox Beta version from Microsoft Store following this particular deploy process
- Read cmake and Kodi documentation
- Provide a fix which fixe the issue, is secure for Kodi users, UWP specific, is compatible with "manual" Kodi compilation and compatible "magic windows" compilation documented in Kodi repo
- Follow Kodi Pull Request process and collaborate with Kodi team
If the solution coded here appear to be "easy to be done", it's because i did a great job :)
Behind this "pull request", i did a lot of job.

**Review by wsnipex — APPROVED** [review]

File: cmake/scripts/windowsstore/Install.cmake [hunk 1]
```diff
@@ -0,0 +1,10 @@
+# Fix UWP addons security issue caused by empty __init__.py Python Lib files packaged with Kodi
+set(uwp_pythonlibinit_filepattern "${DEPENDENCIES_DIR}/bin/Python/Lib/__init__.py")
+file(GLOB_RECURSE uwp_pythonlibinit_foundfiles "${uwp_pythonlibinit_filepattern}")
+foreach(uwp_pythonlibinit_file ${uwp_pythonlibinit_foundfiles})
+    file(SIZE "${uwp_pythonlibinit_file}" uwp_pythonlibinit_filesize)
+    if(${uwp_pythonlibinit_filesize} EQUAL 0)
+        message("Adding hash comment character in the following empty file: ${uwp_pythonlibinit_file}")
+        file(APPEND ${uwp_pythonlibinit_file} "#")
+    endif()
+endforeach()
```
Comment (thexai) [inline cmake/scripts/windowsstore/Install.cmake:10]: One minor:

Please add a blank line at the end

**Comment by matthuisman** [comment]
so, i downloaded the 18.9 uwp and looked at it's python (python2 btw)
there are still 0 byte __init__ files, but not on often used libraries (like urllib in python3)

image

email/mime seems like the only one that may caused an issue
Pretty sure https://kodi.wiki/view/Add-on:WatchedList uses email/mime (or one of it's dependencies do)

If we could find someone with 18.9 retail on xbox to try, they could try an add-on with
`from email.mime import image` in it and see if it fails

**Comment by thexai** [comment]
> Also, seeing so many of the init have a comment in them, I wonder if the 0 byte files are an oversight and should PR against upstream Python to add the comments?

+1
In any case, this PR is still valid as a security measure in case new 0-byte files are somehow introduced in the future.

File: cmake/scripts/windowsstore/Install.cmake (hunk 1 above)
Comment (Chicopower) [reply cmake/scripts/windowsstore/Install.cmake]: @thexai I applied your recommandation. Thx 👍

**Review by thexai — APPROVED** [review]
Approved because has ben tested in Retail mode and it is known 100% that works.

Any other potential solution (even if is "better" on the paper) is not applicable if is not tested first on Xbox in Retail mode.

**Comment by keithah** [comment]
Hey @Chicopower thanks so much for your contributions to Kodi! If you're interested, we'd love to have us join the dev team on slack. Please email me keith-at-kodi.tv so I can send you an invite and get you setup!

**Comment by thexai** [comment]
Congratulations on finding an ingenious solution that does not require modifying Python sources and especially for all the work behind it.

It's actually a specific UWP/Xbox issue so a specific workaround is fine.

**Comment by thexai** [comment]
@Chicopower 
Now that it is merged into master, we need to do the same on the Matrix branch. It is about opening a new PR and in the description it is enough to put "backport of #20501". Your branch must be created from the Matrix branch and the commit can be the same (git cherry-pick) or a new one with the same changes. This is usually done by the same person who is the author of the original commit, although it is not essential.

**Comment by LoneFenris** [comment]
> If we could find someone with 18.9 retail on xbox to try, they could try an add-on with `from email.mime import image` in it and see if it fails
> 
> that would at least tell us if 18.9 had the same issue but just wasn't aware.

If still helpful, I still have 18.9 on my One X in retail mode. If you know an addon I can test, I can give it a shot.

**Comment by matthuisman** [comment]
@LoneFenris 
Here is a quick test add-on I wrote:
https://f.mjh.nz/xbox/plugin.program.xbox.test.zip

On Kodi 18.9 - only the Mime test should ERROR (if it does indeed have same issue as 19)
The urllib and cryptodome test should be OK.

**Comment by LoneFenris** [comment]
> 
> 
> @LoneFenris Here is a quick test add-on I wrote: https://f.mjh.nz/xbox/plugin.program.xbox.test.zip
> 
> On Kodi 18.9 - only the Mime test should ERROR (if it does indeed have same issue as 19) The urllib and cryptodome test should be OK.

Oddly enough, the cryptodome test ERRORs, but the Mime and urllib test OK.

**Comment by matthuisman** [comment]
@LoneFenr
Could you please try again with new version from same link.
my cryptodome test was broken.

Please see if they all pass now

**Comment by LoneFenris** [comment]
@matthuisman Yes -- all three tests now report OK.

**Comment by matthuisman** [comment]
Ok, well then I suspect this same issue was **not** present in 18.9
If 18.9 had the same problem, the mime test should have failed.

Basing this on:
downloading: http://mirrors.kodi.tv/releases/windows/UWP/Kodi-18.9-Leia_18.9.0.0_x64.appx
and confirming System > Python >  Lib > email > mime > __init__.py is a 0 byte file

But then I guess the issue could actually be that Python2 doesn't require X permission on __init__.py & Python3 does.

Anyway - not a huge deal.
Just interesting!

**Pull request merged**